File: fanpage-backend/src/main/java/com/example/fanpagebackend/modules/admin/service/AdminService.java
package com.example.fanpagebackend.modules.admin.service;

import com.example.fanpagebackend.modules.auth.dto.request.*;
import com.example.fanpagebackend.modules.user.dto.request.*;
import com.example.fanpagebackend.modules.post.dto.request.*;
import com.example.fanpagebackend.modules.comment.dto.request.*;
import com.example.fanpagebackend.modules.chat.dto.request.*;
import com.example.fanpagebackend.modules.admin.dto.request.*;
import com.example.fanpagebackend.modules.report.dto.request.*;
import com.example.fanpagebackend.modules.auth.dto.response.*;
import com.example.fanpagebackend.modules.user.dto.response.*;
import com.example.fanpagebackend.modules.post.dto.response.*;
import com.example.fanpagebackend.modules.comment.dto.response.*;
import com.example.fanpagebackend.modules.like.dto.response.*;
import com.example.fanpagebackend.modules.friend.dto.response.*;
import com.example.fanpagebackend.modules.chat.dto.response.*;
import com.example.fanpagebackend.modules.admin.dto.response.*;
import com.example.fanpagebackend.common.dto.response.*;


import java.util.List;
import com.example.fanpagebackend.common.PageResponse;
import com.example.fanpagebackend.modules.admin.dto.request.AdminAssignStaffRoleRequest;
import com.example.fanpagebackend.modules.admin.dto.request.AdminCreateRoleRequest;
import com.example.fanpagebackend.modules.admin.dto.request.AdminResolveReportRequest;
import com.example.fanpagebackend.modules.admin.dto.request.AdminUpdateRoleRequest;
import com.example.fanpagebackend.modules.admin.dto.response.AdminCommentResponse;
import com.example.fanpagebackend.modules.admin.dto.response.AdminDashboardStatsResponse;
import com.example.fanpagebackend.modules.admin.dto.response.AdminPermissionResponse;
import com.example.fanpagebackend.modules.admin.dto.response.AdminPostResponse;
import com.example.fanpagebackend.modules.admin.dto.response.AdminReportResponse;
import com.example.fanpagebackend.modules.admin.dto.response.AdminRoleResponse;
import com.example.fanpagebackend.modules.admin.dto.response.AdminUserResponse;
import com.example.fanpagebackend.modules.accountdeletion.dto.response.AccountDeletionRequestResponse;


public interface AdminService {
    AdminDashboardStatsResponse getDashboardStats();
    PageResponse<AdminUserResponse> getUsers(int page, int size);
    PageResponse<AdminUserResponse> getStaff(int page, int size);
    AdminUserResponse changeUserRole(Long userId, String role);
    AdminUserResponse assignStaffRole(Long userId, AdminAssignStaffRoleRequest request);
    AdminUserResponse removeStaff(Long userId);
    AdminUserResponse lockUser(Long userId, AdminLockUserRequest request);
    AdminUserResponse unlockUser(Long userId);
    PageResponse<AccountDeletionRequestResponse> getAccountDeletionRequests(String status, int page, int size);
    AccountDeletionRequestResponse approveAccountDeletionRequest(Long requestId);
    AccountDeletionRequestResponse rejectAccountDeletionRequest(Long requestId, String adminNote);
    List<AdminPermissionResponse> getPermissions();
    PageResponse<AdminRoleResponse> getRoles(int page, int size);
    AdminRoleResponse createRole(AdminCreateRoleRequest request);
    AdminRoleResponse updateRole(Long roleId, AdminUpdateRoleRequest request);
    void deleteRole(Long roleId);
    PageResponse<AdminPostResponse> getPosts(int page, int size);
    AdminPostResponse updatePostStatus(Long postId, String status);
    void deletePost(Long postId);
    PageResponse<AdminCommentResponse> getComments(int page, int size);
    void deleteComment(Long commentId);
    PageResponse<AdminReportResponse> getReports(int page, int size);
    
    
    AdminReportResponse resolveReportWithAction(Long reportId, String action);
}

File: fanpage-frontend/src/services/adminService.js
import apiClient from './apiClient';

const unwrap = (response) => {
  if (response?.data && typeof response.data === 'object' && 'data' in response.data) {
    return response.data.data;
  }
  return response?.data ?? response;
};

const toItems = (data) => Array.isArray(data) ? data : (Array.isArray(data?.items) ? data.items : []);
const pageQuery = (page = 0, size = 1000) => `?page=${page}&size=${size}`;

const adminService = {
  async getStats() { return unwrap(await apiClient.get('/admin/dashboard/stats')); },
  async getUsers(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/users${pageQuery(page, size)}`))); },
  async getStaff(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/staff${pageQuery(page, size)}`))); },
  async getPermissions() { return unwrap(await apiClient.get('/admin/permissions')); },
  async getRoles(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/roles${pageQuery(page, size)}`))); },
  async createRole(payload) { return unwrap(await apiClient.post('/admin/roles', payload)); },
  async updateRole(roleId, payload) { return unwrap(await apiClient.put(`/admin/roles/${roleId}`, payload)); },
  async deleteRole(roleId) { return unwrap(await apiClient.delete(`/admin/roles/${roleId}`)); },
  async assignStaffRole(userId, roleId) { return unwrap(await apiClient.patch(`/admin/staff/${userId}/staff-role`, { roleId })); },
  async removeStaff(userId) { return unwrap(await apiClient.delete(`/admin/staff/${userId}`)); },
  async changeUserRole(userId, role) { return unwrap(await apiClient.patch(`/admin/users/${userId}/role`, { role })); },
  async lockUser(userId, payload) { return unwrap(await apiClient.patch(`/admin/users/${userId}/lock`, payload)); },
  async unlockUser(userId) { return unwrap(await apiClient.patch(`/admin/users/${userId}/unlock`, {})); },
  async getAccountDeletionRequests(status = 'ALL', page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/account-deletion-requests${pageQuery(page, size)}&status=${encodeURIComponent(status)}`))); },
  async approveAccountDeletionRequest(requestId) { return unwrap(await apiClient.patch(`/admin/account-deletion-requests/${requestId}/approve`, {})); },
  async rejectAccountDeletionRequest(requestId, adminNote = '') { return unwrap(await apiClient.patch(`/admin/account-deletion-requests/${requestId}/reject`, { adminNote })); },
  async getPosts(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/posts${pageQuery(page, size)}`))); },
  async updatePostStatus(postId, status) { return unwrap(await apiClient.patch(`/admin/posts/${postId}/status`, { status })); },
  async deletePost(postId, reason = '') { return unwrap(await apiClient.delete(`/admin/posts/${postId}`, { reason })); },
  async getComments(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/comments${pageQuery(page, size)}`))); },
  async deleteComment(commentId) { return unwrap(await apiClient.delete(`/admin/comments/${commentId}`)); },
  async getReports(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/reports${pageQuery(page, size)}`))); },
  async resolveReport(reportId, payload) { return unwrap(await apiClient.patch(`/admin/reports/${reportId}`, payload)); },
  async getBannedWords(page = 0, size = 1000) { return toItems(unwrap(await apiClient.get(`/admin/banned-words${pageQuery(page, size)}`))); },
  async addBannedWord(payload) { return unwrap(await apiClient.post('/admin/banned-words', payload)); },
  async deleteBannedWord(id) { return unwrap(await apiClient.delete(`/admin/banned-words/${id}`)); }
};

export default adminService;
